rust-src/ed25519_hd_key_derivation/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+use hmac::{Hmac, Mac};
+use regex::Regex;
+use sha2::Sha512;
+
+const ED25519_CURVE: &[u8; 12] = b"ed25519 seed";
+const HARDENED_OFFSET: u32 = 0x80000000;
+
+pub struct HdKeys {
+    pub key:    [u8; 32],
+    chain_code: [u8; 32],
+}
+
+fn get_master_key_from_seed(seed: &[u8]) -> HdKeys {
+    let mut mac = Hmac::<Sha512>::new_from_slice(ED25519_CURVE).unwrap();
+    mac.update(&seed);
+    let i = mac.finalize().into_bytes();
+    let mut il = [0u8; 32];
+    il.clone_from_slice(&i[0..32]);
+    let mut ir = [0u8; 32];
+    ir.clone_from_slice(&i[32..64]);
+
+    HdKeys {
+        key:        il,
+        chain_code: ir,
+    }
+}
+
+fn ckd_priv(keys: HdKeys, index: u32) -> HdKeys {
+    if index & HARDENED_OFFSET == 0 {
+        panic!("Received a non-hardened index!");
+    }
+
+    let mut data = vec![0u8];
+    data.extend_from_slice(&keys.key);
+    data.extend_from_slice(&index.to_be_bytes());
+
+    let mut mac = Hmac::<Sha512>::new_from_slice(&keys.chain_code).unwrap();
+    mac.update(&data);
+    let i = mac.finalize().into_bytes();
+    let mut il = [0u8; 32];
+    il.clone_from_slice(&i[0..32]);
+    let mut ir = [0u8; 32];
+    ir.clone_from_slice(&i[32..64]);
+
+    HdKeys {
+        key:        il,
+        chain_code: ir,
+    }
+}
+
+fn is_valid_path(path: &str) -> bool {
+    let path_regex = Regex::new("^m(/[0-9]+')+$").unwrap();
```

Comment:
I'd prefer not to bring in regex just for this, but I see that it is not that straightforward to implement this check.
I suppose this is not used in any context with malicious input where we'd have to worry about the performance of regex matching and its pathologies.

Reply:
Is this format of the path something we came up with or is it standard. The use of `'` is a bit weird.

Reply:
For our own use cases the `path` will always be valid. I think checking it "manually" will be quite messy.

The format of the path is the standard way of denoting a hardened path, so we cannot get rid of that. I assume it is because all the numbers become quite unmanageable to read if they are all greater than `2^31`.

Reply:
BIP-0032 defines "hardened" and "non-hardened" keys, where for non-hardened ones, you can derive child keys from parent keys, hardened keys don't allow that. Adding a `'` (and setting the highest order bit) indicates that you want a hardened key. We only use hardened keys.

Reply:
Ok. It's a weird format, but if it is standard I have no further comments or objections.